### ISSUE TITLE
Add timestamps to Python Console / kernel logs

### ISSUE THREAD

**Issue by wesm** [posit-dev/positron#4275]
The console logging from Python contains no timestamps, so it is difficult to assess the timeline for requests and their completions. I suggest that we add timestamps so the log so that we can understand the exact timeline

**Comment by petetronic** [comment]
We'd like this sooner than later to help when debugging issues.

**Comment by juliasilge** [comment]
I looked into this a little bit today, and we are not including timestamps on purpose. 😮 

In #1589 we added a new kind of output channel that doesn't add timestamps, and the log for the "Console" channel uses this:

https://github.com/posit-dev/positron/blob/0fe56040a6bb2b8fa547576e10fb2ad1bd96a955/extensions/jupyter-adapter/src/JupyterKernel.ts#L320

I am fuzzy on why we don't want the timestamps here. I will quote @lionel- when considering using the regular output channel log mechanism:

> We can't do the latter option however because that automatically adds timestamps (which we already add elsewhere)

However, I don't think the R and Python console logs have clear timestamps in a traditional sense.

### Example from R

```
[Positron] SEND rpc_reply: OK
[Positron] RECV stream from iopub: {"header":{"msg_id":"be56134e-27fd-4e20-855c-8c9106468539","session":"3d4e54e5-0395-482b-8ece-bfa581e1d967","username":"kernel","date":"2024-09-25T20:03:58.583186+00:00","msg_type":"stream","version":"5.3"},"parent_header":{"msg_id":"fragment-76563419-0cd9-4d2c-9828-95ebdb04f229","session":"c16b75563c8d11eedf6b157a30a905e8","username":"juliasilge","date":"2024-09-25T20:03:53.529Z","msg_type":"execute_request","version":"5.0"},"metadata":{},"content":{"name":"stdout","text":"\u001b[33mℹ\u001b[39m PR \u001b]8;;https://github.com/posit-dev/positron-docs/pull/2\u0007posit-dev/positron-docs/#2\u001b]8;;\u0007 is unmerged, we will not delete the remote branch\n  \u001b[34m\"origin/test-deploy-preview\"\u001b[39m.\n"},"buffers":[]}
[Positron] RECV execute_reply from shell: {"header":{"msg_id":"98f96ca8-105b-43af-9cb1-dec33fbd6ee2","session":"3d4e54e5-0395-482b-8ece-bfa581e1d967","username":"kernel","date":"2024-09-25T20:03:58.584436+00:00","msg_type":"execute_reply","version":"5.3"},"parent_header":{"msg_id":"fragment-76563419-0cd9-4d2c-9828-95ebdb04f229","session":"c16b75563c8d11eedf6b157a30a905e8","username":"juliasilge","date":"2024-09-25T20:03:53.529Z","msg_type":"execute_request","version":"5.0"},"metadata":{},"content":{"status":"ok","execution_count":1,"user_expressions":{}},"buffers":[]}
[Positron] RECV comm_msg/positron-ui-r-1-2c6b8d3a/undefined from iopub: {"header":{"msg_id":"756e44c6-c629-4dca-9704-2d97a47d48ef","session":"3d4e54e5-0395-482b-8ece-bfa581e1d967","username":"kernel","date":"2024-09-25T20:03:58.586197+00:00","msg_type":"comm_msg","version":"5.3"},"parent_header":{},"metadata":{},"content":{"comm_id":"positron-ui-r-1-2c6b8d3a","data":{"method":"busy","params":{"busy":false}}},"buffers":[]}
[Positron] RECV comm_msg/positron-ui-r-1-2c6b8d3a/undefined from iopub: {"header":{"msg_id":"17187a6a-8134-48fd-bd96-5f7e19b4aea2","session":"3d4e54e5-0395-482b-8ece-bfa581e1d967","username":"kernel","date":"2024-09-25T20:03:58.586326+00:00","msg_type":"comm_msg","version":"5.3"},"parent_header":{},"metadata":{},"content":{"comm_id":"positron-ui-r-1-2c6b8d3a","data":{"method":"prompt_state","params":{"input_prompt":"> ","continuation_prompt":"+ "}}},"buffers":[]}
[Positron] R kernel status changed: busy => idle
[Positron] RECV status/idle from iopub: {"header":{"msg_id":"62dd52be-31ee-4cec-876a-0049b5077815","session":"3d4e54e5-0395-482b-8ece-bfa581e1d967","username":"kernel","date":"2024-09-25T20:03:58.586368+00:00","msg_type":"status","version":"5.3"},"parent_header":{"msg_id":"fragment-76563419-0cd9-4d2c-9828-95ebdb04f229","session":"c16b75563c8d11eedf6b157a30a905e8","username":"juliasilge","date":"2024-09-25T20:03:53.529Z","msg_type":"execute_request","version":"5.0"},"metadata":{},"content":{"execution_state":"idle"},"buffers":[]}
[Positron] SEND heartbeat with timeout of 30 seconds
```

A good comparison is what the logs look like in, say, the "Positron R Extension" output channel:

```
2024-09-25 16:35:30.671 [info] Creating Positron R 4.4.0 language client (port 51229)
2024-09-25 16:35:30.671 [info] ARK (R 4.4.0) language client state changed uninitialized => starting
2024-09-25 16:35:30.972 [info] ARK (R 4.4.0) language client init successful
2024-09-25 16:35:30.972 [info] ARK (R 4.4.0) language client state changed starting => running
2024-09-25 16:35:37.702 [info] Log level: Debug
2024-09-25 16:35:37.702 [info] Refreshing testthat status
2024-09-25 16:35:37.705 [info] Not working in an R package
2024-09-25 16:35:37.705 [info] Context key 'testthatIsConfigured' is 'false'
2024-09-25 16:35:37.705 [info] Context key 'testthatHasTests' is 'false'
```

### Example from Python

```
[Positron] SEND execute_request: OK
[Positron] Python kernel status changed: idle => busy
[Positron] RECV status/busy from iopub: {"header":{"msg_id":"a6f6be11-463b0b1410bf666452a5d840_33408_43","msg_type":"status","username":"juliasilge","session":"a6f6be11-463b0b1410bf666452a5d840","date":"2024-09-25T22:35:46.329024Z","version":"5.3"},"parent_header":{"msg_id":"fragment-6f49eae5-500b-45c1-9959-46ce69717287","msg_type":"execute_request","version":"5.0","date":"2024-09-25T22:35:46.327000Z","session":"13647a35d1ce5c647081edbf81c834db","username":"juliasilge"},"metadata":{},"content":{"execution_state":"busy"},"buffers":[]}
[Positron] RECV execute_input from iopub: {"header":{"msg_id":"a6f6be11-463b0b1410bf666452a5d840_33408_44","msg_type":"execute_input","username":"juliasilge","session":"a6f6be11-463b0b1410bf666452a5d840","date":"2024-09-25T22:35:46.329292Z","version":"5.3"},"parent_header":{"msg_id":"fragment-6f49eae5-500b-45c1-9959-46ce69717287","msg_type":"execute_request","version":"5.0","date":"2024-09-25T22:35:46.327000Z","session":"13647a35d1ce5c647081edbf81c834db","username":"juliasilge"},"metadata":{},"content":{"code":"alt.Chart(source).mark_bar().encode(\n    x='a',\n    y='b'\n)\n\n","execution_count":4},"buffers":[]}
[Positron] RECV execute_result from iopub: {"header":{"msg_id":"a6f6be11-463b0b1410bf666452a5d840_33408_45","msg_type":"execute_result","username":"juliasilge","session":"a6f6be11-463b0b1410bf666452a5d840","date":"2024-09-25T22:35:46.336076Z","version":"5.3"},"parent_header":{"msg_id":"fragment-6f49eae5-500b-45c1-9959-46ce69717287","msg_type":"execute_request","version":"5.0","date":"2024-09-25T22:35:46.327000Z","session":"13647a35d1ce5c647081edbf81c834db","username":"juliasilge"},"metadata":{},"content":{"data":{"text/html":"\n<style>\n  #altair-viz-441823f17bdf46c8889d060a792c697c.vega-embed {\n    width: 100%;\n    display: flex;\n  }\n\n  #altair-viz-441823f17bdf46c8889d060a792c697c.vega-embed details,\n  #altair-viz-441823f17bdf46c8889d060a792c697c.vega-embed details summary {\n    position: relative;\n  }\n</style>\n<div id=\"altair-viz-441823f17bdf46c8889d060a792c697c\"></div>\n<script type=\"text/javascript\">\n  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n  (function(spec, embedOpt){\n    let outputDiv = document.currentScript.previousElementSibling;\n    if (outputDiv.id !== \"altair-viz-441823f17bdf46c8889d060a792c697c\") {\n      outputDiv = document.getElementById(\"altair-viz-441823f17bdf46c8889d060a792c697c\");\n    }\n    const paths = {\n      \"vega\": \"https://cdn.jsdelivr.net/npm/vega@5?noext\",\n      \"vega-lib\": \"https://cdn.jsdelivr.net/npm/vega-lib?noext\",\n      \"vega-lite\": \"https://cdn.jsdelivr.net/npm/vega-lite@5.17.0?noext\",\n      \"vega-embed\": \"https://cdn.jsdelivr.net/npm/vega-embed@6?noext\",\n    };\n\n    function maybeLoadScript(lib, version) {\n      var key = `${lib.replace(\"-\", \"\")}_version`;\n      return (VEGA_DEBUG[key] == version) ?\n        Promise.resolve(paths[lib]) :\n        new Promise(function(resolve, reject) {\n          var s = document.createElement('script');\n          document.getElementsByTagName(\"head\")[0].appendChild(s);\n          s.async = true;\n          s.onload = () => {\n            VEGA_DEBUG[key] = version;\n... (truncated)
[Positron] Python kernel status changed: busy => idle
[Positron] RECV status/idle from iopub: {"header":{"msg_id":"a6f6be11-463b0b1410bf666452a5d840_33408_47","msg_type":"status","username":"juliasilge","session":"a6f6be11-463b0b1410bf666452a5d840","date":"2024-09-25T22:35:46.357718Z","version":"5.3"},"parent_header":{"msg_id":"fragment-6f49eae5-500b-45c1-9959-46ce69717287","msg_type":"execute_request","version":"5.0","date":"2024-09-25T22:35:46.327000Z","session":"13647a35d1ce5c647081edbf81c834db","username":"juliasilge"},"metadata":{},"content":{"execution_state":"idle"},"buffers":[]}
[Positron] RECV execute_reply from shell: {"header":{"msg_id":"a6f6be11-463b0b1410bf666452a5d840_33408_46","msg_type":"execute_reply","username":"juliasilge","session":"a6f6be11-463b0b1410bf666452a5d840","date":"2024-09-25T22:35:46.357057Z","version":"5.3"},"parent_header":{"msg_id":"fragment-6f49eae5-500b-45c1-9959-46ce69717287","msg_type":"execute_request","version":"5.0","date":"2024-09-25T22:35:46.327000Z","session":"13647a35d1ce5c647081edbf81c834db","username":"juliasilge"},"metadata":{"started":"2024-09-25T22:35:46.329273Z","dependencies_met":true,"engine":"97649d18-f07d-4c71-9cfa-5328b0ecb02c","status":"ok"},"content":{"status":"ok","execution_count":4,"user_expressions":{},"payload":[]},"buffers":[]}
[Positron] SEND heartbeat with timeout of 30 seconds
```

Are we happy with the decision made back a year ago? Or we do want traditional timestamps (not in the JSON) in these logs?

**Comment by jmcphers** [comment]
In the R logs, some of the lines come from the R log file (prefixed with `[R] `). We get the R logs by having R write to a log file (using Rust's environment logger), then streaming the log to Positron's output channel.

```
[R]   2024-09-25T23:11:49.942301Z  INFO  Received request to open comm: JupyterMessage { zmq_identities: [[0, 128, 0, 65, 167]], header: JupyterHeader { msg_id: "f8d81357f1", session: "r-4562dccf", username: "", date: "2024-09-25T23:11:49.941Z", msg_type: "comm_open", version: "5.3" }, parent_header: None, content: CommOpen { comm_id: "positron-help-r-2-5bb88ab2", target_name: "positron.help", data: Object {} } }
[R]     at crates/amalthea/src/socket/shell.rs:309
[R] 
```

These lines not only have their own timestamp but that timestamp _differs_ from what an auto timestamp would show (since the timestamp shows the time in the R process when the log was emitted, not the time Positron received the message).

So my recollection is that we turned off auto timestamp in order to avoid double/confusing timestamps with the R logs.

No reason we couldn't selectively re-enable timestamps for Python, or do something more nuanced for R.

**Comment by juliasilge** [comment]
At the beginning of the Console logs (both R and Python), it _does_ have the timestamp:

![Screenshot 2024-09-25 at 7 50 11 PM](https://github.com/user-attachments/assets/08e50d54-6d14-4dda-b99c-f2f144aa76b1)

Looks like the last entry in both cases to use the timestamp (still a regular VS Code log???) is the last time we hit this:

https://github.com/posit-dev/positron/blob/0fe56040a6bb2b8fa547576e10fb2ad1bd96a955/extensions/jupyter-adapter/src/JupyterSocket.ts#L241-L242

By the time we are logging these, it no longer has the timestamps:

https://github.com/posit-dev/positron/blob/0fe56040a6bb2b8fa547576e10fb2ad1bd96a955/extensions/jupyter-adapter/src/JupyterSocket.ts#L202-L203

**Comment by juliasilge** [comment]
I spent a little bit of time and want to highlight a few things I observe:

- The "Console" output channel is, in most situations, dominated by the Jupyter messages and other log info that we generate through very normal methods, not the special ark/Python logs that we have to stream through in a tricky way.
- We know whether these Jupyter-message-and-friends messages should be error, warn, info, etc but we are not able to include this info in a structured way that plays nicely with VS Code infrastructure and affordances.
- We know the timestamps for these Jupyter-message-and-friends messages but we are not able to include that either.

I'd like to propose that we **separate** the logs we can generate in a typical VS Code extension way from the ones we have to stream in. It would mean this extension would be generating yet another output channel, but I think that would be worth it since there isn't much info in the kernel logs in most use cases.

It would mean we would lose the interleaved logs we get right now, with, for example, ark logs interleaved with Jupyter message logs, but I will be honest I have found those confusing.

Thoughts?
